Added another overloaded WiFiSTAClass::begin() function that provides…

jpswensen opened the pull request:
… an easy way of creating a WPA2 Enterprise connections.

-----------
## Summary
The examples demonstrate how to create a WPA2 Enterprise connection, but it requires using various direct esp_idf functions. This patch is intended to create another overloaded version of the WiFi.begin() function that allows a user to create a WPA2 Enterprise connection in much the same way as different kinds of connections.

My only question for the core maintainers is whether I should leave those #ifdef's in there. I added them so that it was easy to disable all the code I added via defines from my platformio.ini file, but they technically aren't necessary.

## Impact
This should make it easier for novice users to create WPA2 Enterprise connections. For my university, I didn't need a root certificate or the client certificate or client key, so I haven't been able to debug those scenarios, but I built the begin functions to allow any one of those to be used, if needed.

I can confirm that eduroam-style WPA2 Enterprise networks that only require authentication with a username and password works as expected.

## Related links
N/A

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=6398) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=6398) before we can accept your contribution.<br/><hr/>**John P. Swensen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/arduino-esp32?pullRequest=6398) it.</sub>

me-no-dev commented on libraries/WiFi/src/WiFiGeneric.h:
```diff
@@ -31,6 +31,12 @@
 
+#ifdef ENABLE_WPA2_AUTHENTICATION
+#include <WiFiClient.h>
+#include <WiFiClientSecure.h>
```
why are those two included? Just `esp_wpa2.h` is enough, no?

jpswensen replied:
I don't think that esp_wpa2.h is enough. If you look just above the new WiFiSTAClass::begin() function that I created, I declared a static WiFiClientSecure. This is necessary for negotiating the various types of cert processes that are sometimes needed with WPA2 Enterprise. esp_wpa2.h does not include WiFiClientSecure.h, as the former is a esp-idf header and the latter is an arduino-espressif32 header.

I could/should arguably move these into the WiFiSTA.cpp file (and I don't think I actually need WiFiClient.h) to prevent unnecessary #includes, however.

me-no-dev replied:
yes please :)

LinkedInprofile replied:
MSVALIDATE.01 ¤ "Meta" <tag>
-----[\HTTPS://BITCOINNEWS.COM----]

LinkedInprofile replied:
What is was is s someone added their Business camera sytem to my computer... now i have identity issues

LinkedInprofile replied:
"Screwballs"tied his camera system to my computer... now i cannot access any of my own stuff
![20220405_174156](https://user-images.githubusercontent.com/103056448/161875533-5f234cb4-cb9d-4b62-9255-347a0a60323e.jpg)

me-no-dev commented on libraries/WiFi/src/WiFiGeneric.h:
```diff
@@ -31,6 +31,12 @@
 #include "esp_smartconfig.h"
 #include "wifi_provisioning/manager.h"
 
+#ifdef ENABLE_WPA2_AUTHENTICATION
```
does it need to be disabled? Would it not work for all chips/cases? Having it guarded like that would require some way to enable this function.

jpswensen replied:
I had this guard in there because I wasn't sure if there was any shared code between the WiFi portions of the Arduino cores for ESP8266 and ESP32. AFAIK based on fairly recent testing, I don't think the WPA2 Enterprise authentication works for ESP8266, so I was thinking ahead to try and make it capable of being guarded easily. I can remove those, if you want me to.

me-no-dev replied:
I see the client nowhere being used other than to set some certs that yu provide as arguments to the function. You should be able to safely skip those and add them to the sketch. BTW could you please provide an example of the use?

LinkedInprofile replied:
I have a company profile on linkedin that can be used to identify me currently, a IBM is a computer of the guy that is stealing my identity.

jpswensen commented:
OK. I think my latest two commits both fix those issues and fixed another issue I realized I had. Since I don't have a WiFi system available that requires cert files, I don't have a way of testing those. I can test on my local WPA2 Enterprise system that requires PEAP identity, username, and password (but not cert files). However, as I was double checking everything I realized that I had mistakenly been working off an example that was showing interaction with a secure server using a secure client, instead of showing the cert-file WPA2 Enterprise authentication.

So, I went back to the base esp-idf example that demonstrates the cert-file WPA2 Enterprise authentication and made fixes to ensure this is now correct (I think).  I still don't have a way of testing this, but I think it should be correct (assuming the esp-idf SDK example works).

I also modified the WiFiClientEnterprise.ino example in the WiFi library to use this new method and added a whole bunch of comments to explain the three most common modes of use that a typical user would see (TLS+certs, TTLS/PEAP+no_certs, TTLS/PEAP+certs).

LinkedInprofile commented on libraries/WiFi/src/WiFiGeneric.h:
```diff
@@ -31,6 +31,12 @@
+
 ESP_EVENT_DECLARE_BASE(ARDUINO_EVENTS);
 
 typedef enum {
```
✔🛫🛫🛫✔🛫🛫🔗 👋👋👋

D﻿ VERSION:3. 0﻿ N:﻿ FN:﻿ EMAIL;TYPE=PREF:hello@247atmformula. com﻿ END:VCARD﻿
hello@247atmformula.com📁📂👋

LinkedInprofile replied:
[Linkedinprofile](https://bitcoinnewsamerica.com)

LinkedInprofile replied:
[default.PDF](https://github.com/espressif/arduino-esp3libraries/WiFi/src/WiFiGeneric.h
2/files/8422799/default.PDF)

LinkedInprofile reviewed (commented):
[Pantyshac93](© https://780013295261.signin.aws.amazon.com/cons )

LinkedInprofile reviewed (commented):
[Pantyshac93](© https://780013295261.signin.aws.amazon.com/cons )
